File: android/app/src/main/java/com/pangomicro/zpfeng/cylinder_qrcode/Arc.java
package com.pangomicro.zpfeng.cylinder_qrcode;

import org.opencv.core.Mat;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Collections;
import java.util.Set;

import Jama.*;
/**
 * Created by zczx1 on 2016/4/11.
 */
public class Arc {
    public Arc() {
        pts = new ArrayList<>();
    }

    public boolean setPts(List<Point> srcPts, boolean is_filled_between, boolean sort_X_or_Y) {
        if (srcPts.size() < 2)
            return false;
        if (sort_X_or_Y)
            Collections.sort(srcPts, new sortLessX());
        else
            Collections.sort(srcPts, new sortLessY());
        pts = new ArrayList<>();
        if (is_filled_between) {
            if (srcPts.size() == 1)
                pts.add(srcPts.get(0));
            else
                for (int i = 0; i < srcPts.size()-1; ++i) {
                    insertBetween(pts, srcPts.get(i), srcPts.get(i+1));
                }
        } else {
            pts.addAll(srcPts);
        }
        if (sort_X_or_Y) {
            startX = pts.get(0);
            endX = pts.get(pts.size()-1);
            double ys = pts.get(0).y, ye = pts.get(0).y;
            int ysi = 0, yei = 0;
            for (int i = 1; i < pts.size(); ++i) {
                if (ys > pts.get(i).y) {
                    ysi = i;
                    ys = pts.get(i).y;
                }
                if (ye < pts.get(i).y) {
                    yei = i;
                    ye = pts.get(i).y;
                }
            }
            startY = pts.get(ysi);
            endY = pts.get(yei);
        } else {
            startY = pts.get(0);
            endY = pts.get(pts.size()-1);
            double xs = pts.get(0).x, xe = pts.get(0).x;
            int xsi = 0, xei = 0;
            for (int i = 1; i < pts.size(); ++i) {
                if (xs > pts.get(i).x) {
                    xsi = i;
                    xs = pts.get(i).x;
                }
                if (xe < pts.get(i).x) {
                    xei = i;
                    xe = pts.get(i).x;
                }
            }
            startX = pts.get(xsi);
            endX = pts.get(xei);
        }
        center_point = new Point(0, 0);
        center_point.x += startX.x;
        center_point.x += endX.x;
        center_point.x += startY.x;
        center_point.x += endY.x;
        center_point.y += startX.y;
        center_point.y += endX.y;
        center_point.y += startY.y;
        center_point.y += endY.y;
        center_point.x /= 4;
        center_point.y /= 4;
        return true;
    }

    private void insertBetween(List<Point> insertPts, Point pt1, Point pt2) {
//        if (pt1 == pt2) {
//            insertPts.add(pt1);
//            return;
//        }
        Set<Point> hashSet = new HashSet<>();
        List<Point> newList = new ArrayList<>();
        if (hashSet.add(pt1))
            newList.add(pt1);
        Point tmp = new Point(-1, -1);
        Point ntmp;
        if (pt1.x == pt2.x) {
            tmp.x = pt1.x;
            if (pt1.y > pt2.y) {
                for (int i = (int)(pt1.y-1); i > pt2.y; --i) {
                    tmp.y = i;
                    ntmp = new Point(tmp.x, tmp.y);
                    if (hashSet.add(ntmp))
                        newList.add(ntmp);
                }
            } else if (pt1.y < pt2.y) {
                for (int i = (int)(pt1.y+1); i < pt2.y; ++i) {
                    tmp.y = i;
                    ntmp = new Point(tmp.x, tmp.y);
                    if (hashSet.add(ntmp))
                        newList.add(ntmp);
                }
            }
        } else {
            double a = (pt1.y-pt2.y)/(pt1.x-pt2.x);
            double b = pt1.y-pt1.x*a;
            if (pt1.x > pt2.x) {
                for (int i = (int)(pt1.x-1); i > pt2.x; --i) {
                    tmp.x = i;
                    tmp.y = a*i+b;
                    ntmp = new Point(tmp.x, tmp.y);
                    if (hashSet.add(ntmp))
                        newList.add(ntmp);
                }
            } else if (pt1.x < pt2.x) {
                for (int i = (int)(pt1.x+1); i < pt2.x; ++i) {
                    tmp.x = i;
                    tmp.y = a*i+b;
                    ntmp = new Point(tmp.x, tmp.y);
                    if (hashSet.add(ntmp))
                        newList.add(ntmp);
                }
            }
        }
        if (hashSet.add(pt2))
            newList.add(pt2);
        for (int i = 0; i < newList.size(); ++i)
            insertPts.add(newList.get(i));
    }

    public void copyPtsTo(List<Point> dst) {
        if (pts.isEmpty())
            return;
        for (int i = 0; i < pts.size(); ++i)
            dst.add(pts.get(i));
    }

    public List<Point> getPts() {
        return pts;
    }

    public void drawArc(Mat drawPic, Scalar color) {
        drawAllPts(drawPic, color);
    }

    public void drawAllPts(Mat drawPic, Scalar color) {
        for (int i = 0; i < pts.size(); ++i) {
            Imgproc.circle(drawPic, pts.get(i), 2, color);
        }
    }

    public void drawParabola(Mat drawPic, Vec3d para, Scalar color, Size s,
                             boolean is_Y, Point sp, Point ep) {
        Point p1, p2, s1, s2;
        p1 = new Point(-1, -1);
        p2 = new Point(-1, -1);
        s1 = sp.clone();
        s2 = ep.clone();
        if (is_Y) {
            if (s1.x > s2.x) {
                Point tmp;
                tmp = s1;
                s1 = s2;
                s2 = tmp;
            }
            p1.x = s1.x;
            p1.y = para.getA()*s1.x*s1.x+para.getB()*s1.x+para.getC();
            double ty;
            for (double x = p1.x; x <= s2.x; ++x) {
                ty = para.getA()*x*x+para.getB()*x+para.getC();
                p2.x = x;
                p2.y = ty;
                if (ty >= 0 && ty < s.height && p1.y >=0 && p1.y < s.height) {
                    Imgproc.line(drawPic, p1, p2, color, 3);
                }
                p1.x = p2.x;
                p1.y = p2.y;
            }
        } else {
            if (s1.y > s2.y) {
                Point tmp;
                tmp = s1;
                s1 = s2;
                s2 = tmp;
            }
            p1.y = s1.y;
            p1.x = para.getA()*s1.y*s1.y+para.getB()*s1.y+para.getC();
            double tx;
            for (double y = p1.y; y <= s2.y; ++y) {
                tx = para.getA()*y*y+para.getB()*y+para.getC();
                p2.x = tx;
                p2.y = y;
                if (tx >= 0 && tx < s.width && p1.x >=0 && p1.x < s.width) {
                    Imgproc.line(drawPic, p1, p2, color, 3);
                }
                p1.x = p2.x;
                p1.y = p2.y;
            }
        }
    }

    public Vec3d lineFittingX() throws Exception{
        return lineFitting(this.pts, false);
    }

    public Vec3d lineFittingY() throws Exception{
        return lineFitting(this.pts, true);
    }

    public Vec3d lineFitting(List<Point> srcPts, boolean is_Y) throws Exception{
        Matrix A, At, b;
        A = new Matrix(srcPts.size(), 2);
        b = new Matrix(srcPts.size(), 1);
        if (is_Y) {
            for (int i = 0; i < srcPts.size(); ++i) {
                A.set(i, 0, srcPts.get(i).x);
                A.set(i, 1, 1);
                b.set(i, 0, srcPts.get(i).y);
            }
        } else {
            for (int i = 0; i < srcPts.size(); ++i) {
                A.set(i, 0, srcPts.get(i).y);
                A.set(i, 1, 1);
                b.set(i, 0, srcPts.get(i).x);
            }
        }
        At = A.transpose();
        Matrix As = At.times(A);
        Matrix invA;
//        try {
            invA = As.inverse();
//        } catch (RuntimeException e) {
//            invA = new Matrix(2, 2, 0);
//        }
        Matrix res;
        res = invA.times(At);
        res = res.times(b);
        return new Vec3d(0, res.get(0, 0), res.get(1, 0));
    }

    public Vec3d parabolaFittingX(boolean is_Rectify, Point rectifyPoint)  throws Exception{
        return parabolaFitting(this.pts, false, is_Rectify, rectifyPoint);
    }

    public Vec3d parabolaFittingY(boolean is_Rectify, Point rectifyPoint)  throws Exception{
        return parabolaFitting(this.pts, true, is_Rectify, rectifyPoint);
    }

    public Vec3d parabolaFitting(List<Point> srcPts, boolean is_Y,
                                       boolean is_Rectify, Point rectifyPoint)  throws Exception{
        Matrix A, At, b;
        if (is_Rectify) {
            A = new Matrix(srcPts.size()+1, 3);
            b = new Matrix(srcPts.size()+1, 1);
        } else {
            A = new Matrix(srcPts.size(), 3);
            b = new Matrix(srcPts.size(), 1);
        }
        if (is_Y) {
            for (int i = 0; i < srcPts.size(); ++i) {
                A.set(i, 0, (srcPts.get(i).x)*(srcPts.get(i).x));
                A.set(i, 1, srcPts.get(i).x);
                A.set(i, 2, 1);
                b.set(i, 0, srcPts.get(i).y);
            }
            if (is_Rectify) {
                A.set(srcPts.size(), 0, rectifyPoint.x*rectifyPoint.x);
                A.set(srcPts.size(), 1, rectifyPoint.x);
                A.set(srcPts.size(), 2, 1);
                b.set(srcPts.size(), 0, rectifyPoint.y);
            }
        } else {
            for (int i = 0; i < srcPts.size(); ++i) {
                A.set(i, 0, (srcPts.get(i).y)*(srcPts.get(i).y));
                A.set(i, 1, srcPts.get(i).y);
                A.set(i, 2, 1);
                b.set(i, 0, srcPts.get(i).x);
            }
            if (is_Rectify) {
                A.set(srcPts.size(), 0, rectifyPoint.y*rectifyPoint.y);
                A.set(srcPts.size(), 1, rectifyPoint.y);
                A.set(srcPts.size(), 2, 1);
                b.set(srcPts.size(), 0, rectifyPoint.x);
            }
        }
        At = A.transpose();
        Matrix As = At.times(A);
        Matrix invA;
//        try {
            invA = As.inverse();
//        } catch (RuntimeException e) {
//            invA = new Matrix(3, 3, 0);
//        }
        Matrix res;
        res = invA.times(At);
        res = res.times(b);
        return new Vec3d(res.get(0, 0), res.get(1, 0), res.get(2, 0));
    }

    public Point getStartEndXY(int i) {
        int r = i%4;
        Point res = new Point(-1, -1);
        switch (r) {
            case 0:
                res = startX;
                break;
            case 1:
                res = endX;
                break;
            case 2:
                res = startY;
                break;
            case 3:
                res = endY;
                break;
        }
        return res;
    }

    private List<Point> pts;

    public boolean isHas_two() {
        return has_two;
    }

    public void setHas_two(boolean has_two) {
        this.has_two = has_two;
    }

    private boolean has_two;
    private Point startX, endX, startY, endY;

    public Point getCenter_point() {
        return center_point;
    }

    private Point center_point;
}

class sortLessX implements Comparator<Point> {

    @Override
    public int compare(Point lhs, Point rhs) {
        if (lhs.x < rhs.x)
            return -1;
        else if (lhs.x > rhs.x)
            return 1;
        return 0;
    }
}

class sortLessY implements Comparator<Point> {

    @Override
    public int compare(Point lhs, Point rhs) {
        if (lhs.y < rhs.y)
            return -1;
        else if (lhs.y > rhs.y)
            return 1;
        return 0;
    }
}

File: src/arc.cpp
#include "arc.h"
#include <math.h>

const double ePI = 3.14159265358979323846264338327950288419716939937510;

bool qr_cylinder::LessX(const Point &lp, const Point &rp)
{
    return lp.x < rp.x;
}

bool qr_cylinder::LessY(const Point &lp, const Point &rp)
{
    return lp.y < rp.y;
}

qr_cylinder::Arc::Arc()
    :_has_two(false)
{
}

/* sort_X_or_Y
 * true: sort(pts.begin(), pts.end(), lessX);
 * false: sort(pts.begin(), pts.end(), lessY);
 */
void qr_cylinder::Arc::SetPts(vector<Point> & arc_pts, bool is_filled_between, bool sort_x_or_y)
{
    if (arc_pts.empty())
        return;
    if (sort_x_or_y)
        sort(arc_pts.begin(), arc_pts.end(), LessX);
    else
        sort(arc_pts.begin(), arc_pts.end(), LessY);
    if (!_pts.empty())
        _pts.clear();
    if (is_filled_between)
        for (size_t i = 0; i < (arc_pts.size()-1); ++i) {
            InsertBetween(arc_pts[i], arc_pts[i+1]);
        }
    else {
        _pts.assign(arc_pts.begin(), arc_pts.end());
    }
    if (sort_x_or_y) {
        _start_x = _pts[0];
        _end_x = _pts[_pts.size()-1];
        double ys = _pts[0].y, ye = _pts[0].y;
        int ysi = 0, yei = 0;
        for (size_t i = 1; i < _pts.size(); ++i) {
            if (ys > _pts[i].y) {
                ysi = i;
                ys = _pts[i].y;
            }
            if (ye < _pts[i].y) {
                yei = i;
                ye = _pts[i].y;
            }
        }
        _start_y = _pts[ysi];
        _end_y = _pts[yei];
    } else {
        _start_y = _pts[0];
        _end_y = _pts[_pts.size()-1];
        double xs = _pts[0].x, xe = _pts[0].x;
        int xsi = 0, xei = 0;
        for (size_t i = 0; i < _pts.size(); ++i) {
            if (xs > _pts[i].x) {
                xsi = i;
                xs = _pts[i].x;
            }
            if (xe < _pts[i].x) {
                xei = i;
                xe = _pts[i].x;
            }
        }
        _start_x = _pts[xsi];
        _end_x = _pts[xei];
    }
}

const vector<Point>& qr_cylinder::Arc::GetPts()
{
    return _pts;
}

Point qr_cylinder::Arc::GetStartEndXY(int i)
{
    int r = i%4;
    Point res(-1, -1);
    switch (r) {
    case 0:
        res = _start_x;
        break;
    case 1:
        res = _end_x;
        break;
    case 2:
        res = _start_y;
        break;
    case 3:
        res = _end_y;
        break;
    }
    return res;
}

void qr_cylinder::Arc::SetTwo(bool b)
{
    _has_two = b;
}

bool qr_cylinder::Arc::HasTwo()
{
    return _has_two;
}

void qr_cylinder::Arc::DrawAllPts(Mat &src, const Scalar &color)
{
    for (size_t i = 0; i < _pts.size(); ++i) {
        circle(src, _pts[i], 1, color);
    }
}

void qr_cylinder::Arc::InsertBetween(const Point &pt1, const Point &pt2)
{
    _pts.push_back(pt1);

    //push points between them to pts
    Point tmp;
    if (pt1.x == pt2.x) {
        tmp.x = pt1.x;
        if (pt1.y > pt2.y) {
            for (int i = (pt1.y-1); i > pt2.y; --i) {
                tmp.y = i;
                _pts.push_back(tmp);
            }
        } else if (pt1.y < pt2.y) {
            for (int i = (pt1.y+1); i < pt2.y; ++i) {
                tmp.y = i;
                _pts.push_back(tmp);
            }
        }
    } else {
        double a = (pt1.y-pt2.y)/(pt1.x-pt2.x);
        double b = pt1.y-pt1.x*a;
        if (pt1.x > pt2.x) {
            for (int i = (pt1.x-1); i > pt2.x; --i) {
                tmp.x = i;
                tmp.y = a*i+b;
                _pts.push_back(tmp);
            }
        } else if (pt1.x < pt2.x) {
            for (int i = (pt1.x+1); i < pt2.x; ++i) {
                tmp.x = i;
                tmp.y = a*i+b;
                _pts.push_back(tmp);
            }
        }
    }

    _pts.push_back(pt2);
    //删除重复的点
    _pts.erase(unique(_pts.begin(), _pts.end()), _pts.end());
}

void qr_cylinder::Arc::CopyPtsTo(vector<Point> &dst)
{
    if (_pts.empty())
        return;
    for (size_t i = 0; i < _pts.size(); ++i)
        dst.push_back(_pts[i]);
}

void qr_cylinder::Arc::DrawArc(Mat &src, const Scalar &color)
{
    DrawAllPts(src, color);
}

void qr_cylinder::Arc::DrawParabola(Mat &draw_pic, const Vec3d &para, const Scalar &color, const Size& s,
                       bool is_y,
                       const Point &start_point, const Point &stop_point)
{
    Point p1, p2;
    Point s1 = start_point;
    Point s2 = stop_point;
    if (is_y) {
        if (s1.x > s2.x) {
            Point tmp = s1;
            s1 = s2;
            s2 = tmp;
        }
        p1.x = s1.x;
        p1.y = para[0]*p1.x*p1.x + para[1]*p1.x + para[2];
        double ty;
        for (int x = s1.x; x <= s2.x; ++x) {
            ty = para[0]*x*x + para[1]*x + para[2];
            p2.x = x;
            p2.y = ty;
            if (ty >= 0 && ty < s.height && p1.y >=0 && p1.y < s.height) {
                line(draw_pic, p1, p2, color);
            }
            p1.x = p2.x;
            p1.y = p2.y;
        }
    } else {
        if (s1.y > s2.y) {
            Point tmp = s1;
            s1 = s2;
            s2 = tmp;
        }
        p1.y = s1.y;
        p1.x = para[0]*p1.y*p1.y + para[1]*p1.y + para[2];
        double tx;
        for (int y = s1.y; y <= s2.y; ++y) {
            tx = para[0]*y*y + para[1]*y + para[2];
            p2.x = tx;
            p2.y = y;
            if (tx >= 0 && tx < s.width && p1.x >=0 && p1.x < s.width) {
                line(draw_pic, p1, p2, color);
            }
            p1.x = p2.x;
            p1.y = p2.y;
        }
    }

}

Vec3d qr_cylinder::Arc::ParabolaFittingY(bool is_rectify, Point rectify_point)
{
    return ParabolaFitting(_pts, true, is_rectify, rectify_point);
}


Vec3d qr_cylinder::Arc::ParabolaFittingX(bool is_rectify, Point rectify_point)
{
    return ParabolaFitting(_pts, false, is_rectify, rectify_point);
}

/*
 * 最小二乘抛物线拟合有两种情况
 * 1. 用y=a*x^2+b*x+c来拟合，求[a b c]三个参数,此时is_Y = true;
 * 2. 用x=a*y^2+b*y+c来拟合，求[a b c]三个参数，此时is_Y = false;
 */
Vec3d qr_cylinder::Arc::ParabolaFitting(const vector<Point> &src_pts, bool is_y, bool is_rectify, Point rp)
{
    Mat A, A_transpose;
    Mat b;
    if (!is_rectify) {
        A.create(src_pts.size(), 3, CV_64FC1);
        b.create(src_pts.size(), 1, CV_64FC1);
    } else {
        A.create(src_pts.size()+1, 3, CV_64FC1);
        b.create(src_pts.size()+1, 1, CV_64FC1);
    }
    if (is_y) {
        for (size_t i = 0; i < src_pts.size(); ++i) {
            A.at<double>(i, 0) = src_pts[i].x*src_pts[i].x;
            A.at<double>(i, 1) = src_pts[i].x;
            A.at<double>(i, 2) = 1;
            b.at<double>(i, 0) = src_pts[i].y;
        }
        if (is_rectify) {
            A.at<double>(src_pts.size(), 0) = rp.x*rp.x;
            A.at<double>(src_pts.size(), 1) = rp.x;
            A.at<double>(src_pts.size(), 2) = 1;
            b.at<double>(src_pts.size(), 0) = rp.y;
        }
    } else {
        for (size_t i = 0; i < src_pts.size(); ++i) {
            A.at<double>(i, 0) = src_pts[i].y*src_pts[i].y;
            A.at<double>(i, 1) = src_pts[i].y;
            A.at<double>(i, 2) = 1;
            b.at<double>(i, 0) = src_pts[i].x;
        }
        if (is_rectify) {
            A.at<double>(src_pts.size(), 0) = rp.y*rp.y;
            A.at<double>(src_pts.size(), 1) = rp.y;
            A.at<double>(src_pts.size(), 2) = 1;
            b.at<double>(src_pts.size(), 0) = rp.x;
        }
    }
    transpose(A, A_transpose);
    Mat A_square = A_transpose*A;
    Mat A_square_inverse;
    invert(A_square, A_square_inverse);
    Mat res = A_square_inverse*A_transpose*b;
    Vec3d r;
    r[0] = res.at<double>(0, 0);
    r[1] = res.at<double>(1, 0);
    r[2] = res.at<double>(2, 0);
    return r;
}

//为了和抛物线拟合兼容，也是使用Vec3d作为返回值
//只是此时二次项为0
Vec3d qr_cylinder::Arc::LineFittingY()
{
    return LineFitting(_pts, true);
}

Vec3d qr_cylinder::Arc::LineFittingX()
{
    return LineFitting(_pts, false);
}

/*
 * 最小二乘直线拟合有两种情况
 * 1. 用y=a*x+b来拟合，求[0 a b]中两个参数,此时is_Y = true;
 * 2. 用x=a*y+b来拟合，求[0 a b]中两个参数，此时is_Y = false;
 */
Vec3d qr_cylinder::Arc::LineFitting(const vector<Point> &src_pts, bool is_y)
{
    Mat A, A_transpose;
    A.create(src_pts.size(), 2, CV_64FC1);
    Mat b;
    b.create(src_pts.size(), 1, CV_64FC1);
    if (is_y) {
        for (size_t i = 0; i < src_pts.size(); ++i) {
            A.at<double>(i, 0) = src_pts[i].x;
            A.at<double>(i, 1) = 1;
            b.at<double>(i, 0) = src_pts[i].y;
        }
    } else {
        for (size_t i = 0; i < src_pts.size(); ++i) {
            A.at<double>(i, 0) = src_pts[i].y;
            A.at<double>(i, 1) = 1;
            b.at<double>(i, 0) = src_pts[i].x;
        }
    }
    transpose(A, A_transpose);
    Mat A_square = A_transpose*A;
    Mat A_square_inverse;
    invert(A_square, A_square_inverse);
    Mat res = A_square_inverse*A_transpose*b;
    Vec3d r;
    r[0] = 0;
    r[1] = res.at<double>(0, 0);
    r[2] = res.at<double>(1, 0);
    return r;
}

double qr_cylinder::Arc::CalParabolaFittingErrorX(const Vec3d &para)
{
    return CalParabolaFittingError(para, _pts, false);
}

double qr_cylinder::Arc::CalParabolaFittingErrorY(const Vec3d &para)
{
    return CalParabolaFittingError(para, _pts, true);
}

/*
 * para: Vec3d [a b c]
 * is_Y
 *     true: y-a*x^2-b*x-c
 */
double qr_cylinder::Arc::CalParabolaFittingError(const Vec3d &para, const vector<Point> &src_pts, bool is_y)
{
    double ans = 0;
    if (is_y)
        for (size_t i = 0; i < src_pts.size(); ++i)
            ans += (src_pts[i].y-(src_pts[i].x*src_pts[i].x*para[0]
                    +src_pts[i].x*para[1]+para[2]));
    else
        for (size_t i = 0; i < src_pts.size(); ++i)
            ans += (src_pts[i].x-(src_pts[i].y*src_pts[i].y*para[0]
                    +src_pts[i].y*para[1]+para[2]));
    return ans;
}

double qr_cylinder::Arc::CalLineFittingErrorY(const Vec2d &para)
{
    return CalLineFittingError(para, _pts, true);
}

double qr_cylinder::Arc::CalLineFittingErrorX(const Vec2d &para)
{
    return CalLineFittingError(para, _pts, false);
}

double qr_cylinder::Arc::CalLineFittingError(const Vec2d &para, const vector<Point> &src_pts, bool is_y)
{
    double ans = 0;
    if (is_y)
        for (size_t i = 0; i < _pts.size(); ++i) {
            ans += (src_pts[i].y-(src_pts[i].x*para[0]+para[1]));
        }
    else
        for (size_t i = 0; i < _pts.size(); ++i) {
            ans += (src_pts[i].x-(src_pts[i].y*para[0]+para[1]));
        }
    return ans;
}

void getMeanXY(const vector<Point> &pts, double &mean_x, double &mean_y)
{
    mean_x = 0;
    mean_y = 0;
    for (size_t i = 0; i < pts.size(); ++i) {
        mean_x += (static_cast<double>(pts[i].x)/pts.size());
        mean_y += (static_cast<double>(pts[i].y)/pts.size());
    }
}

void getCenterXY(const vector<Point> &pts, double &sx, double &sy)
{
    int max_x, min_x, max_y, min_y;
    max_x = pts[0].x;
    min_x = pts[0].x;
    max_y = pts[0].y;
    min_y = pts[0].y;
    for (size_t i = 1; i < pts.size(); ++i) {
        if (max_x < pts[i].x)
            max_x = pts[i].x;
        if (min_x > pts[i].x)
            min_x = pts[i].x;
        if (max_y < pts[i].y)
            max_y = pts[i].y;
        if (min_y > pts[i].y)
            min_y = pts[i].y;
    }
    sx = (max_x-min_x)/2.0;
    sy = (max_y-min_y)/2.0;
}


